src/program_state.inl

```diff
@@ -844,6 +849,163 @@ public:
+
+			if (amd_comgr_destroy_metadata(max_flat_workgroup_size)
+                != AMD_COMGR_STATUS_SUCCESS)
+                continue;
+        }
+
+        if (amd_comgr_destroy_metadata(kernels_md) != AMD_COMGR_STATUS_SUCCESS
+            || amd_comgr_destroy_metadata(metadata) != AMD_COMGR_STATUS_SUCCESS)
+            return;
+
+        amd_comgr_release_data(dataIn);
+    }
+
+    static
+    void read_kernattribute_metadata_v2(
+        const std::string& kernels_md,
+        std::size_t dx,
+        std::unordered_map<
+            std::string,std::size_t>& kernattrib) {
+        do {
+            dx = kernels_md.find("Name:", dx);
+
+            if (dx == std::string::npos) break;
+
+            static constexpr decltype(kernels_md.size()) name_sz{5};
+            dx = kernels_md.find_first_not_of(" '", dx + name_sz);
+
+            auto fn =
+                kernels_md.substr(dx, kernels_md.find_first_of("'\n", dx) - dx);
+            dx += fn.size();
+
+            std::size_t m = kernels_md.find("MaxFlatWorkGroupSize:",dx);
+            auto n = kernels_md.substr(m+21, kernels_md.find_first_of("'\n", m) - m - 21);
+            size_t block_size = std::stoul(n);
+            kernattrib[fn] = block_size;
+
+        } while (true);
+    }
+    static
+    void read_kernattribute_metadata(
+        const std::string& blob,
+        std::unordered_map<
+            std::string,
+            std::size_t>& kernattrib)
+    {
+        std::istringstream istr{blob};
+        ELFIO::elfio reader;
+
+        if (!reader.load(istr)) return;
+
+        // TODO: this is inefficient.
+        auto it = find_section_if(reader, [](const ELFIO::section* x) {
+            return x->get_type() == SHT_NOTE;
+        });
+
+        if (!it) return;
+
+        const ELFIO::note_section_accessor acc{reader, it};
+        auto n{acc.get_notes_num()};
+        while (n--) {
+            ELFIO::Elf_Word type{};
+            std::string name{};
+            void* desc{};
+            ELFIO::Elf_Word desc_size{};
+
+            acc.get_note(n, type, name, desc, desc_size);
+
+            if (name == "AMDGPU") {
+                return read_kernattribute_metadata_v3(blob, kernattrib);
+            }
+            if (name != "AMD") continue; // TODO: switch to using NT_AMD_AMDGPU_HSA_METADATA.
+
+            std::string tmp{
+                static_cast<char*>(desc), static_cast<char*>(desc) + desc_size};
+
+            auto dx = tmp.find("Kernels:");
+
+            if (dx == std::string::npos) continue;
+
+            return read_kernattribute_metadata_v2(tmp, dx + 8u, kernattrib); // Skip "Kernels:".
+        }
+    }
+
+    const std::unordered_map<std::string, std::size_t>& get_kernattribute() {
+
+        std::call_once(kernattrib.first, [this]() {
+            for (auto&& name_and_isa_blobs : get_code_object_blobs()) {
+                for (auto&& isa_blobs : name_and_isa_blobs.second) {
+                    for (auto&& blob : isa_blobs.second) {
```

Comment:
This three-nested for-loops looks daunting, even though it is called once. Anyway to optimize this?